### DISCUSSION TITLE
Mermaid doesn't work

### DISCUSSION THREAD

**Discussion by nicfab** [adityatelange/hugo-PaperMod#850] [Q&A]
Hi!
Starting from the file `single.html`, which is under `themes/PaperMod/layouts/_default`, I edited it adding the following Mermaid code
```
{{ if .Page.Store.Get "hasMermaid" }}
  <script src="https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js"></script>
  <script>
    mermaid.initialize({ startOnLoad: true });
  </script>
{{ end }}
```
I retrieved that code from the Hugo website, here https://gohugo.io/content-management/diagrams/#mermaid-diagrams
I put the modified file `single.html` in `/layouts/_default/`, leaving the original `single.html` under `themes/PaperMod/layouts/_default`.

Now the files `single.hml` under `/layouts/_default/` has the following lines:

```
{{- define "main" }}

<article class="post-single">
  <header class="post-header">
    {{ partial "breadcrumbs.html" . }}
    <h1 class="post-title">
      {{ .Title }}
      {{- if .Draft }}<sup><span class="entry-isdraft">&nbsp;&nbsp;[draft]</span></sup>{{- end }}
    </h1>
    {{- if .Description }}
    <div class="post-description">
      {{ .Description }}
    </div>
    {{- end }}
    {{- if not (.Param "hideMeta") }}
    <div class="post-meta">
      {{- partial "post_meta.html" . -}}
      {{- partial "translation_list.html" . -}}
      {{- partial "edit_post.html" . -}}
      {{- partial "post_canonical.html" . -}}
    </div>
    {{- end }}
  </header>
  {{- $isHidden := .Params.cover.hidden | default .Site.Params.cover.hiddenInSingle | default .Site.Params.cover.hidden }}
  {{- partial "cover.html" (dict "cxt" . "IsHome" false "isHidden" $isHidden) }}
  {{- if (.Param "ShowToc") }}
  {{- partial "toc.html" . }}
  {{- end }}

  {{- if .Content }}
  <div class="post-content">
    {{- if not (.Param "disableAnchoredHeadings") }}
    {{- partial "anchored_headings.html" .Content -}}
    {{- else }}{{ .Content }}{{ end }}
  </div>
  {{ if .Page.Store.Get "hasMermaid" }}
  <script src="https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js"></script>
  <script>
    mermaid.initialize({ startOnLoad: true });
  </script>
  {{ end }}
  {{- end }}
  
  <footer class="post-footer">
    {{- if .Params.tags }}
    <ul class="post-tags">
      {{- range ($.GetTerms "tags") }}
      <li><a href="{{ .Permalink }}">{{ .LinkTitle }}</a></li>
      {{- end }}
    </ul>
    {{- end }}
    {{- if (.Param "ShowPostNavLinks") }}
    {{- partial "post_nav_links.html" . }}
    {{- end }}
    {{- if (and .Site.Params.ShowShareButtons (ne .Params.disableShare true)) }}
    {{- partial "share_icons.html" . -}}
    {{- end }}
  </footer>

  {{- if (.Param "comments") }}
  {{- partial "comments.html" . }}
  {{- end }}
</article>

{{- end }}{{/* end main */}}
```

Unfortunately, Mermaid doesn't work.
Where was I wrong?

**Comment by adityatelange** [comment]
Try implementating like it is implemented in this commit
 
https://github.com/bee-san/hugo-PaperMod/commit/c5c9ffc3006f5fb499b2a8c66b94bdd3800433eb

**Reply by adityatelange** [reply]
Oh wait the doc mentions this path
`layouts/_default/_markup/render-codeblock-mermaid.html`

**Reply by nicfab** [reply]
> Try implementating like it is implemented in this commit
> 
> [bee-san@c5c9ffc](https://github.com/bee-san/hugo-PaperMod/commit/c5c9ffc3006f5fb499b2a8c66b94bdd3800433eb)

You suggest seeing the commit you mentioned.
Should I intervene o the files under the theme folders, or should I copy them under /layout?

**Reply by nicfab** [reply]
> Oh wait the doc mentions this path
> `layouts/_default/_markup/render-codeblock-mermaid.html`

What do you mean?
I am sorry, but it is not clear to me.

**Reply by adityatelange** [reply]
```html
<div class="mermaid">
  {{- .Inner | safeHTML }}
</div>
{{ .Page.Store.Set "hasMermaid" true }}
```

Should be in `layouts/_default/_markup/render-codeblock-mermaid.html`

and 

```html
{{ if .Page.Store.Get "hasMermaid" }}
  <script src="https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js"></script>
  <script>
    mermaid.initialize({ startOnLoad: true });
  </script>
{{ end }}
```

should go in `layouts/partials/extend_footer.html`

**Reply by nicfab** [reply]
Thank you @adityatelange 
I did it so.

I copied the following lines:

```
{{ if .Page.Store.Get "hasMermaid" }}
  <script src="https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js"></script>
  <script>
    mermaid.initialize({ startOnLoad: true });
  </script>
{{ end }}
```

into the `extended_head.html`, and now it seems to be working.
My question is: "When I will update the theme, will the file `extend_head.html` be overwritten?

**Reply by asphaltbuffet** [reply]
> into the `extended_head.html`, and now it seems to be working.

**Thank you for this!**

I was having the same problems (also had been referencing the same examples as noted above). Using your tip, I found that if I put the shortcode to insert script loading in `extend_footer.html` it won't work, but if it's in `extend_head.html` it works fine.

With a little debug output, it seems that `{{ if .Page.Store.Get "hasMermaid" }}` is never evaluated to true when in the footer. I don't have a good enough understanding to know why that placement makes a difference though.

**Reply by yatharthb97** [reply]
I tried adding it to both `extend_footer.html` and `extend_head.html` files, but it didn't work for me. It works if I modify the `single.html` file in the theme's `layout/_default` folder. Which is the method prescribed here: https://gohugo.io/content-management/diagrams/#mermaid-diagrams  
("And then include this snippet at the bottom of the content template (below .Content):")


But this is not convenient when hosting on GitHub via GitHub Actions. Maybe a direct integration of `mermaid` in the theme is better? (with a switch parameter like `mermaid: true` ?

**Reply by asphaltbuffet** [reply]
> But this is not convenient when hosting on GitHub via GitHub Actions.

None of those should affect the ability to use this with Github actions. Are you able to check the rendered source to ensure that the files are getting rendered as you expect?

**Reply by baozaolaowu** [reply]
Neither extended_head.html, extended_footer.html nor single.html work for my site.
I am prompted with the following error.
`<.Page.Store.Get>: can't evaluate field Store in type page.Page`
<img width="861" alt="image" src="https://user-images.githubusercontent.com/9347231/210293609-2b84afdc-c34e-4303-b0b7-34b58bf7bc86.png">

**Reply by ArkaprabhaChakraborty** [reply]
> Neither extended_head.html, extended_footer.html nor single.html work for my site. I am prompted with the following error. `<.Page.Store.Get>: can't evaluate field Store in type page.Page` <img alt="image" width="861" src="https://user-images.githubusercontent.com/9347231/210293609-2b84afdc-c34e-4303-b0b7-34b58bf7bc86.png">

Me too man!!! Anyone knows why?

**Reply by willihay** [reply]
> > Neither extended_head.html, extended_footer.html nor single.html work for my site. I am prompted with the following error. `<.Page.Store.Get>: can't evaluate field Store in type page.Page` <img alt="image" width="861" src="https://user-images.githubusercontent.com/9347231/210293609-2b84afdc-c34e-4303-b0b7-34b58bf7bc86.png">
> 
> Me too man!!! Anyone knows why?

Yes, I just had to troubleshoot this. If you see the error message "`.Page.Store.Get>: can't evaluate field Store in type page.Page`", then you need to upgrade your version of Hugo. The `Page.Store` field was added in **0.93**.

**Reply by baozaolaowu** [reply]
I succeeded with this method.
In the first step, Add the following code to `layouts/partials/extend_head.html`.
```
<script src="https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js"></script>
<script>
const config = {
    startOnLoad:true,
    theme: 'forest',
    themeVariables: {
        lineColor: "#fafafa"  
    },
    flowchart: {
        useMaxWidth:false,
        htmlLabels:true
        }
};
mermaid.initialize(config);
window.onload = () => {
    window.mermaid.init(undefined, document.querySelectorAll('.language-mermaid'));
}
</script>
````
In the last step, the `guessSyntax` of the goldmark in the config.yaml file has to be set to `false` or deleted for Mermaid to take effect.
<img width="439" alt="image" src="https://user-images.githubusercontent.com/9347231/220865398-53de709f-2818-41f6-905c-497ea0b448b1.png">

**Comment by howardjohn** [comment]
Here is how I did it:

```diff
diff --git a/layouts/partials/extend_head.html b/layouts/partials/extend_head.html
new file mode 100644
index 0000000..59db893
--- /dev/null
+++ b/layouts/partials/extend_head.html
@@ -0,0 +1,9 @@
+{{ if .Page.Store.Get "needMermaid" -}}
+<script type="module">
+  import mermaid from 'https://cdn.jsdelivr.net/npm/mermaid@10/dist/mermaid.esm.min.mjs';
+  mermaid.initialize({
+    startOnLoad: true,
+    'theme': 'base',
+  });
+</script>
+{{ end -}}
diff --git a/layouts/shortcodes/mermaid.html b/layouts/shortcodes/mermaid.html
new file mode 100644
index 0000000..c00edf5
--- /dev/null
+++ b/layouts/shortcodes/mermaid.html
@@ -0,0 +1,2 @@
+{{ .Page.Store.Set "needMermaid" true }}
+<div class="mermaid" align="center">{{ safeHTML .Inner }}</div>
```

Then in the content:

```
{{< mermaid >}}
graph LR
  a-->b
{{< /mermaid >}}
```

**Reply by howardjohn** [reply]
With changing based on color scheme (although not until page refresh):

```
{{ if .Page.Store.Get "needMermaid" -}}
<script type="module">
  var theme = 'neutral';
    if (localStorage.getItem("pref-theme") == "dark") {
      theme = 'dark'
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
      theme = 'dark'
    }
  import mermaid from 'https://cdn.jsdelivr.net/npm/mermaid@10/dist/mermaid.esm.min.mjs';
  mermaid.initialize({
    startOnLoad: true,
    'theme': theme,
  });
</script>
{{ end -}}
```

**Comment by nakednous** [comment]
I tried following @howardjohn by editing both `partials` and `shortcodes` in local `layouts` (not those found in `themes` as I plan to implement other shortcodes without affecting the upstream theme) so that they look like this:

`layouts/partials/extend_head.html`:
```html
<!--katex is working -->
{{ if or .Params.math .Site.Params.math }}
{{ partial "math.html" . }}
{{ end }}

<!--mermaid is NOT working -->
{{ if .Page.Store.Get "needMermaid" -}}
<script type="module">
  var theme = 'neutral';
    if (localStorage.getItem("pref-theme") == "dark") {
      theme = 'dark'
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
      theme = 'dark'
    }
  import mermaid from 'https://cdn.jsdelivr.net/npm/mermaid@10/dist/mermaid.esm.min.mjs';
  mermaid.initialize({
    startOnLoad: true,
    'theme': theme,
  });
</script>
{{ end -}}
```

`layouts/shortcodes/mermaid.html `:
```html
{{ .Page.Store.Set "needMermaid" true }}
<div class="mermaid" align="center">{{ safeHTML .Inner }}</div>
```

when including the shortcode within `mypost.md` as:

```md
{{< mermaid >}}
graph LR
  a-->b
{{< /mermaid >}}
```

I got no mermaid diagram rendered (but just a center aligned text `graph LR a-->b`).

Any advice is much appreciated.

**Comment by fguisso** [comment]
Just a tip for anyone searching for toggle theme mode using mermaid, live example [here]():
`layouts/_default/_markup/render-codeblock-mermaid.html`
```golang
<div class="mermaid">
  {{- .Inner | safeHTML }}
</div>
{{ .Page.Store.Set "hasMermaid" true }}
```

`layouts/partials/extend_footer.html`
```golang
{{ if .Page.Store.Get "hasMermaid" }}
<script src="https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js"></script>
<script>
    const elementCode = '.mermaid'
    const loadMermaid = function(theme) {
        
        mermaid.initialize({theme})
        mermaid.init({theme}, document.querySelectorAll(elementCode))
    }
    const saveOriginalData = function(){
        return new Promise((resolve, reject) => {
            try {
                var els = document.querySelectorAll(elementCode),
                    count = els.length;
                els.forEach(element => {
                    element.setAttribute('data-original-code', element.innerHTML)
                    count--
                    if(count == 0){
                        resolve()
                    }
                });
            } catch (error) {
                reject(error)
            }
        })
    }
    const resetProcessed = function(){
        return new Promise((resolve, reject) => {
            try {
                var els = document.querySelectorAll(elementCode),
                    count = els.length;
                els.forEach(element => {
                    if(element.getAttribute('data-original-code') != null){
                        element.removeAttribute('data-processed')
                        element.innerHTML = element.getAttribute('data-original-code')
                    }
                    count--
                    if(count == 0){
                        resolve()
                    }
                });
            } catch (error) {
                reject(error)
            }
        })
    }

    saveOriginalData()
        .catch( console.error )
   
    let statusTheme = localStorage.getItem("pref-theme")
        if (statusTheme == 'dark') {
            resetProcessed()
                .then(loadMermaid('dark'))
                .catch(console.error)
        }
        if (statusTheme == 'light') {
            resetProcessed()
                .then(loadMermaid('neutral'))
                .catch(console.error)
        }
       document.getElementById("theme-toggle").addEventListener("click",()=>{
        resetProcessed()
            .then(loadMermaid(mermaid.mermaidAPI.getConfig().theme == 'neutral' ? 'dark' : 'neutral'))
            .catch(console.error)

    })
</script>
{{ end }}
```
thanks to https://github.com/mermaid-js/mermaid/issues/1945#issuecomment-1661264708

**Comment by syaghoubi00** [comment]
What I came up with based on [Diagrams | Hugo](https://gohugo.io/content-management/diagrams/#mermaid-diagrams)

`layouts/_default/_markup/render-codeblock-mermaid.html`
```html
<pre class="mermaid">
  {{- .Inner | htmlEscape | safeHTML }}
</pre>
{{ .Page.Store.Set "hasMermaid" true }}
```

`layouts/partials/extend_head.html`
```html
{{ if .Store.Get "hasMermaid" }}
<script src="https://cdn.jsdelivr.net/npm/mermaid@11/dist/mermaid.min.js" integrity="sha256-pDvBr9RG+cTMZqxd1F0C6NZeJvxTROwO94f4jW3bb54=" crossorigin="anonymous"></script>
<script>
  mermaid.initialize({ 
    startOnLoad: true,
  });
</script>
{{ end }}
```

Now within pages I don't need to use params or shortcodes.
A codeblock and specify `mermaid` as the language and it renders.

`page.md`

````md
```mermaid
graph TD;
    A-->B;
    A-->C;
    B-->D;
    C-->D;
```
````

---
As an aside, I was messing with [GoAT](https://github.com/bep/goat) diagrams, which `hugo` natively supports. These may be good enough for some use cases. This [site](https://diagon.arthursonzogni.com/) made creating some simple diagrams easy. So that may be something to consider if you don't wan't/need to depend on mermaid.

**Reply by fabiojmendes** [reply]
I was trying to do basically the same thing using `extend_footer.html` instead of `extend_head` and was having all kinds of problems.

Turns out that the footer is using `partialCached` on [baseof.html](https://github.com/adityatelange/hugo-PaperMod/blob/5a4651783fa9159123d947bd3511b355146d4797/layouts/_default/baseof.html#L24), so the conditional include won't work.

**Comment by prajravi** [comment]
I ran into the same issue and built a solution that addresses the common pitfalls people have hit in this thread, including the `partialCached` problem that @fabiojmendes flagged (which breaks any approach that puts conditional logic in `footer.html` or `extend_footer.html`).

My implementation adds three files:

1. A render hook (`layouts/_markup/render-codeblock-mermaid.html`) that converts fenced mermaid code blocks automatically
2. A conditional script block in `baseof.html` (outside the cached footer) that only loads on pages with mermaid diagrams
3. Minimal CSS for the diagram container

It uses a fixed light Mermaid theme with a neutral background, which solves the dark mode readability issue without needing live theme-toggle re-rendering.

Users enable it with one config param:

```yaml
params:
  mermaid:
    source: "https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.esm.min.mjs"
```

No CDN is hardcoded in the theme. Pages without diagrams have zero overhead.

Working branch: https://github.com/prajravi/hugo-PaperMod/tree/feat/mermaid-support

Happy to open a PR if maintainers are interested in including this natively. The repo restricts PRs to collaborators, so posting here first.
